### ISSUE TITLE
Infallible enum <-> primitive conversion. eg enum X { Unknown(u16), A=1, ....

### ISSUE THREAD

**Issue by apps4uco** [illicitonion/num_enum#44]
I realise that what I want requires rust language support first but I was wondering if it could be put on the roadmap...

The relevant issue is
Tracking issue for RFC 2363, "Allow arbitrary enums to have explicit discriminants"
#60553 
https://github.com/rust-lang/rust/issues/60553

My idea is that supported enums are represented by known enum tags and
other values are represented by Unknown(u16) which has the discriminant of the u16 parameter, and that the conversion from and to a primitive u16 is infallible. 

This would be useful for FFI to and from other languages,
and in my case from extensible file formats.

Code:

~~~
use num_enum::FromPrimitive;
#[derive(Debug, Eq, PartialEq, FromPrimitive)]
#[repr(u16)]
pub enum TagEnum {
    Unknown(u16),
    ImageWidth = 0x0100,
...etc
~~~

Code example:
https://gist.github.com/e0ef47efa0bb5b3865cf5ddcc3eae92d

**Comment by illicitonion** [comment]
Thanks for filing the suggestion! I think some way of supporting this could be really useful, possibly under some attribute; I'm imagining something like:

```rust
use num_enum::FromPrimitive;
#[derive(Debug, Eq, PartialEq, FromPrimitive)]
#[repr(u16)]
pub enum TagEnum {
    #[num_enum::catch_all] // or wrapped_value or default_wrapper or something...
    Unknown(u16),
    ImageWidth = 0x0100,
}
```

One interesting question here is around the symmetry between `From` and `Into`. While I think it's reasonable to expect:

```rust
assert_eq!(TagEnum::Unknown(1), TagEnum::from(1))
```

It's less clear to me which of these assertions should be correct: 
```rust
let value: u16 = TagEnum::Unknown(1).into()
assert_eq!(value, 0) // Because 0 is the discriminant of the enum
assert_eq!(value, 1) // Because 1 is the value wrapped in the default wrapper variant
```

While we _could_ just allow deriving `FromPrimitive` and not `IntoPrimitive` in this case, it would be nicer to have a consistent story around them both...

---

Right now if you have a variant which has fields, we error [1]. If instead, we skipped that variant from our codegen'd `match`, and let it fall through to the default match case to return an error (which we could do under very controlled circumstances - I'm imagining "there's exactly one variant with exactly one field with the type of the repr, and maybe require an attribute on it too"), you could write code like this:

```rust
#![feature(arbitrary_enum_discriminant)]

use num_enum::TryFromPrimitive;

#[derive(Debug, Eq, PartialEq, TryFromPrimitive)]
#[repr(u16)]
enum TagEnum {
    Unknown(u16),
    ImageWidth = 0x0100,
}

impl TagEnum {
    fn from_primitive(value: u16) -> Self {
        Self::try_from_primitive(value).unwrap_or_else(|err| Self::Unknown(err.number))
    }
}

fn main() {
    assert_eq!(TagEnum::from_primitive(7), TagEnum::Unknown(7));
    assert_eq!(TagEnum::from_primitive(0x0100), TagEnum::ImageWidth);
}
```

I quite like this from a perspective of keeping the codegen for the matching logic hidden away in the proc_macro, but having the author of the enum type write explicit code for how to handle the unknown case, as it avoids us needing to make decisions about things like how `Into` should behave for these types (and is also a fairly trivial change for us to make). It still puts a little burden on the enum author to handle the conversion, but that handling is pretty minimal, and quite nicely explicit... What do you think?

---

1: We _should_ error nicely, but instead we happen to generate code which doesn't compile, and _that_ causes an error - I'll fix that now!

**Comment by danieleades** [comment]
i'm looking for a feature like this too.

In my case, I have a protocol with a combination of pre-defined and user-defined discriminators. Pre-defined values should be mapped to concrete enum variants, but undefined ones should not be discarded.

**Comment by illicitonion** [comment]
Sorry for not getting back to this sooner, it's been a busy couple of weeks! I'd be happy to review a PR which implemented this functionality, as long as it required an explicit attribute on the catch-all variant to opt in to the behaviour (and let's not overload `default` for this) :) We should still error for data-bearing enums if the attr isn't present, though.

**Comment by mvforell** [comment]
I'm also looking for something like this, and it's great to see that #70 has been merged and released in 0.5.7! However, if I understand correctly, it is currently not possible to use this with custom discriminant values.

Half because I wanted this feature with support for custom discriminant values, half to finally try out writing proc macros, I've cooked up something. It works around this issue by being a function-style proc macro instead of a derive macro.

Using it looks like this:
```rs
use repr_with_fallback::repr_with_fallback;

repr_with_fallback! {
    #[derive(Debug, PartialEq)]
    pub enum Algorithm {
        RSASHA256 = 8,
        RSASHA512 = 10,
        ECDSAP256SHA256 = 13,
        ECDSAP384SHA384 = 14,
        ED25519 = 15,
        Unassigned(u8),
    }
}

fn main() {
    assert_eq!(u8::from(Algorithm::ED25519), 15);
    assert_eq!(Algorithm::from(15), Algorithm::ED25519);

    assert_eq!(u8::from(Algorithm::Unassigned(17)), 17);
    assert_eq!(Algorithm::from(17), Algorithm::Unassigned(17));
}
```

It currently imposes two restrictions on the `enum`:
1. It must have only unit variants, except for exactly one variant with exactly one unnamed field. This is used as the fallback variant, and the type of its field must match the type of the discriminants.
2. Every variant must have a discriminant value provided. This was originally decided to reduce the scope of the implementation, but it is probably doable to implement the C-style handling (i.e. first variant has default discriminant 0, every other variant's default discriminant is the discriminant of the variant before plus 1).

An interesting side effect of the second restriction is that the following currently works:
```rs
repr_with_fallback! {
    pub enum Strings {
        Foo = "static",
        Bar = "string",
        Baz = "slices",
        Spam = "work",
        Eggs = "too",
        Unknown(&'static str),
    }
}

fn main() {
    let s: &'static str = Strings::Foo.into();
    assert_eq!(s, "static");
}
```
So it may actually be desirable to keep this restriction (or make two versions of the macro, one with the restriction where this works, and one without the restriction).

Would you be interested in a PR that implements this? If not, I would publish it as a standalone crate.

**Comment by illicitonion** [comment]
This got implemented in #70 and released in 0.5.7 :)
